src/protocol/prf_sharding/feature_label_dot_product.rs

```diff
@@ -0,0 +1,469 @@
+                ctx.narrow(&Step::DidSourceReceiveAttribution),
+                record_id,
+            ),
+        )
+        .await?;
+
+        let (updated_is_saturated, capped_label) = try_join(
+            or(
+                ctx.narrow(&Step::ComputeSaturatingSum),
+                record_id,
+                &self.is_saturated,
+                &did_source_get_attributed,
+            ),
+            did_source_get_attributed.multiply(
+                &(share_of_one - &self.is_saturated),
+                ctx.narrow(&Step::IsAttributedSourceAndPrevRowNotSaturated),
+                record_id,
+            ),
+        )
+        .await?;
+
+        let unbitpacked_feature_vector = BitDecomposed::decompose(FV::BITS, |i| {
+            input_row.feature_vector.map(|v| Gf2::truncate_from(v[i]))
+        });
+
+        let capped_attributed_feature_vector = compute_capped_feature_vector(
+            ctx.narrow(&Step::ComputedCappedFeatureVector),
+            record_id,
+            &capped_label,
+            &unbitpacked_feature_vector,
+        )
+        .await?;
+
+        self.ever_encountered_a_trigger_event = ever_encountered_a_trigger_event;
+        self.is_saturated = updated_is_saturated;
+
+        Ok(capped_attributed_feature_vector)
+    }
+}
+
+#[derive(Step)]
+pub enum UserNthRowStep {
+    #[dynamic]
+    Row(usize),
+}
+
+impl From<usize> for UserNthRowStep {
+    fn from(v: usize) -> Self {
+        Self::Row(v)
+    }
+}
+
+#[derive(Step)]
+pub(crate) enum Step {
+    BinaryValidator,
+    PrimeFieldValidator,
+    EverEncounteredTriggerEvent,
+    DidSourceReceiveAttribution,
+    ComputeSaturatingSum,
+    IsAttributedSourceAndPrevRowNotSaturated,
+    ComputedCappedFeatureVector,
+    ModulusConvertFeatureVectorBits,
+}
+
+fn compute_histogram_of_users_with_row_count<S>(rows_chunked_by_user: &[Vec<S>]) -> Vec<usize> {
+    let mut output = vec![];
+    for user_rows in rows_chunked_by_user {
+        for j in 0..user_rows.len() {
+            if j >= output.len() {
+                output.push(0);
+            }
+            output[j] += 1;
+        }
+    }
+    output
+}
+
+fn set_up_contexts<C>(root_ctx: &C, histogram: &[usize]) -> Vec<C>
+where
+    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+{
+    let mut context_per_row_depth = Vec::with_capacity(histogram.len());
+    for (row_number, num_users_having_that_row_number) in histogram.iter().enumerate() {
+        if row_number == 0 {
+            // no multiplications needed for each user's row 0. No context needed
+        } else {
+            let ctx_for_row_number = root_ctx
+                .narrow(&UserNthRowStep::from(row_number))
+                .set_total_records(*num_users_having_that_row_number);
+            context_per_row_depth.push(ctx_for_row_number);
+        }
+    }
+    context_per_row_depth
+}
+
+fn chunk_rows_by_user<FV>(
+    input_rows: Vec<PrfShardedIpaInputRow<FV>>,
+) -> Vec<Vec<PrfShardedIpaInputRow<FV>>>
```

Comment:
This is a bit unfortunate.  Is it possible to change this to be stream -> stream?  It's OK if the output stream returns a Vec for each user (those should be small), but this won't be nice as it is.

Reply:
Done!